modules/040-node-manager/hooks/check_nodes_cri.go

```diff
@@ -59,25 +76,99 @@ var _ = sdk.RegisterFunc(&go_hook.HookConfig{
 			},
 			FilterFunc: applyNodesCRIVersionFilter,
 		},
+		{
+			Name:                   nodeGroupSnapName,
+			Kind:                   "NodeGroup",
+			ApiVersion:             "deckhouse.io/v1",
+			WaitForSynchronization: pointer.Bool(false),
+			FilterFunc:             applyNodeGroupCRITypeFilter,
+		},
 	},
 }, discoverNodesCRIVersion)
 
 func applyNodesCRIVersionFilter(obj *unstructured.Unstructured) (go_hook.FilterResult, error) {
-	containerVersion, ok, err := unstructured.NestedString(obj.Object, "status", "nodeInfo", "containerRuntimeVersion")
-	if !ok {
-		return containerUnknownVersion, err
+	var node corev1.Node
+
+	err := sdk.FromUnstructured(obj, &node)
+	if err != nil {
+		return nil, err
+	}
+
+	nodeGroup := node.Labels["node.deckhouse.io/group"]
+	containerRuntimeVersion := node.Status.NodeInfo.ContainerRuntimeVersion
+	kubeletVersion := node.Status.NodeInfo.KubeletVersion
+
+	if containerRuntimeVersion == "" {
+		containerRuntimeVersion = containerUnknownVersion
+	}
+
+	return nodeCRIVersion{
+		NodeGroup:               nodeGroup,
+		ContainerRuntimeVersion: containerRuntimeVersion,
+		KubeletVersion:          kubeletVersion,
+	}, nil
+}
+
+func applyNodeGroupCRITypeFilter(obj *unstructured.Unstructured) (go_hook.FilterResult, error) {
+	var ng ngv1.NodeGroup
+
+	err := sdk.FromUnstructured(obj, &ng)
+	if err != nil {
+		return nil, err
 	}
-	return containerVersion, err
+
+	return nodeGroupCRIType{
+		Name:    ng.GetName(),
+		CriType: ng.Spec.CRI.Type,
+	}, nil
 }
 
 func discoverNodesCRIVersion(input *go_hook.HookInput) error {
-	snap := input.Snapshots[snapName]
-	if len(snap) == 0 {
+	ngSnap := input.Snapshots[nodeGroupSnapName]
+	ngCRITypeMap := make(map[string]string)
+
+	notManagedCriKubeVersion, err := semver.NewVersion(notManagedCriMaxKubeVersion)
+	if err != nil {
+		return err
+	}
+
+	for _, item := range ngSnap {
+		ng := item.(nodeGroupCRIType)
+		ngCRITypeMap[ng.Name] = ng.CriType
+	}
+
+	nSnap := input.Snapshots[nodeSnapName]
+	if len(nSnap) == 0 {
 		return nil
 	}
 
-	for _, s := range snap {
-		if !isContainerdRegexp.MatchString(s.(string)) {
+	for _, item := range nSnap {
+		n := item.(nodeCRIVersion)
+		criType, ok := ngCRITypeMap[n.NodeGroup]
+
+		kubeVersion, err := semver.NewVersion(n.KubeletVersion)
+		if err != nil {
+			return err
+		}
+
+		if isContainerdRegexp.MatchString(n.ContainerRuntimeVersion) {
+			continue
+		}
+
+		// not found NodeGroup CRI Type
+		if !isContainerdRegexp.MatchString(n.ContainerRuntimeVersion) && !ok {
+			requirements.SaveValue(hasNodesOtherThanContainerd, true)
+			return nil
+		}
+
+		// skip if NodeGroup CRI Type == NotManaget and node kube ver < notManagedCriKubeVersion
+		if !isContainerdRegexp.MatchString(n.ContainerRuntimeVersion) &&
+			criType == criTypeNotManaged &&
+			kubeVersion.LessThan(notManagedCriKubeVersion) {
+			continue
+		}
+
+		if !isContainerdRegexp.MatchString(n.ContainerRuntimeVersion) {
```

Comment:
i think we do not need !isContainerdRegexp.MatchString(n.ContainerRuntimeVersion) condition due to we check them above.